hw6/minchuncho/matrix.hpp

```diff
@@ -0,0 +1,90 @@
+//
+//  matrix.hpp
+//  nsd-hw3
+//
+//  Created by MinChunCho on 2022/3/25.
+//
+
+#ifndef matrix_hpp
+#define matrix_hpp
+
+#include <vector>
+#include <iostream>
+#include <algorithm>
+#include <stdexcept>
+#include <pybind11/pybind11.h>
+#include <pybind11/operators.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+#include <mkl.h>
+
+namespace py = pybind11;
+
+class Matrix {
+public:
+    Matrix(size_t const& row, size_t const& col);
+    Matrix(size_t const& row, size_t const& col, std::vector<double> const& vec);
+    Matrix(Matrix const& mat);
+    Matrix& operator=(Matrix const& mat);
+    Matrix(Matrix && mat);
+    Matrix& operator=(Matrix && mat);
+    ~Matrix(){
+        vec_.clear();
+        vec_.shrink_to_fit();
+    }
+    
+    Matrix& operator=(std::vector<double> const& vec);
+    double& operator()(size_t const& i, size_t const& j);
+    double operator()(size_t const& i, size_t const& j) const;
+    bool operator==(Matrix const& mat) const;
+    friend std::ostream& operator<<(std::ostream& os, Matrix mat);
+    friend Matrix multiply_naive(Matrix const& ma, Matrix const& mb);
+    friend Matrix multiply_tile(Matrix const& mat1, Matrix const& mat2, size_t tsize);
+    friend Matrix multiply_mkl(Matrix& ma, Matrix& mb);
+
+    size_t row() const { return row_; }
+    size_t col() const { return col_; }
+    size_t size() const { return row_*col_; }
+    double* addr() { return &vec_[0]; }
+    double get_element(std::pair<size_t, size_t> index);
+    void set_element(std::pair<size_t, size_t> index, const double& item);
+    std::string get_matrix_str();
+
+private:
+    bool bound_check(size_t const& i, size_t const& j);
+
+    size_t row_;
+    size_t col_;
+    std::vector<double> vec_;
+};
+
+Matrix multiply_naive(Matrix const& ma, Matrix const& mb);
+Matrix multiply_tile(Matrix const& mat1, Matrix const& mat2, size_t tsize);
+Matrix multiply_mkl(Matrix& ma, Matrix& mb);
+
+PYBIND11_MODULE(_matrix, m) {
+    py::class_<Matrix>(m, "Matrix")
+        .def(py::init<size_t const&, size_t const&>())
+        .def(py::init<size_t const&, size_t const&, std::vector<double> const&>())
+        .def(py::self == py::self)                              // operator overloading
+        .def_property_readonly("nrow", &Matrix::row)            // the following functions are made for validate.sh
+        .def_property_readonly("ncol", &Matrix::col)
+        .def("__setitem__", &Matrix::set_element)
+        .def("__getitem__", &Matrix::get_element)
+        .def("__repr__", &Matrix::get_matrix_str)
+        .def_property("array", [](Matrix &n) -> py::array_t<double>
+        {
+            return py::array_t<double>(
+                { n.row(), n.col() },                               // Buffer dimensions
+                { sizeof(double) * n.col(), sizeof(double) },       // Strides (in bytes) for each index
+                n.addr(),                                            // Pointer to buffer
+                py::cast(n)
```

Comment:
Hi professor @yungyuc 
I found that without this line, values in `mat.array` cannot be changed in python. However, I thought as long as we assign the pointer to the buffer to `py::array_t`,  the buffer values can be set successfully.

Reply:
This corresponds to the last argument `base`  in  https://github.com/pybind/pybind11/blob/80d452484c5409444b0ec19383faa84bb7a4d351/include/pybind11/numpy.h#L555 :

```cpp
    array(const pybind11::dtype &dt, ShapeContainer shape, StridesContainer strides,
          const void *ptr = nullptr, handle base = handle()) {
```

The argument `base` is the owner of the buffer in line 80 (`n.addr()`).  If you do not provide the correct owner, I am not sure what would pybind11 do.  I think it may depend on the Python code you write.

Do you get an exception or any error messages?

Reply:
So based on `def test_ndarray(self)` (https://github.com/minchuncho/nsdhw-22sp/blob/99fa257895115e0eb9925a83151d3d964464fe4f/hw6/validate.py#L107-L127), `mat` is a C++ object but since we're now asking for a `ndarray`, we need to create a python object based on `mat` as the owner of the buffer.
If the argument `base` is not assigned, maybe `mat.array` is owned by another python object and whatever we do to it, it's not related to `mat`.
(Anyways, this is just my guess and there are still questions to it.)

As for error messages or exception, not really.
The only thing I discovered was that calling `mat.array.fill(0)` **seemingly** does nothing to the buffer. (https://github.com/minchuncho/nsdhw-22sp/blob/99fa257895115e0eb9925a83151d3d964464fe4f/hw6/validate.py#L123)

I will delve into this problem some time later.
Thank you :)))

Reply:
I suspect that the returned array does not have the writable bit set.  Could you check it in the Python side?  https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html

Reply:
I checked the flag `WRITEABLE` of `mat.array` and it is set to be `True`.

But I noticed the difference with another flag `OWNDATA`.
To simply put, first I export matrix buffer from C++ **with** and **without** `py::cast(n)`.
```python
.def_property("array", [](Matrix &n) -> py::array_t<double>
{
    return py::array_t<double>(
        { n.row(), n.col() },                               // Buffer dimensions
        { sizeof(double) * n.col(), sizeof(double) },       // Strides (in bytes) for each index
        n.addr(),                                            // Pointer to buffer
        py::cast(n)
    );
}, nullptr)
.def_property("array2", [](Matrix &n) -> py::array_t<double>
{
    return py::array_t<double>(
        { n.row(), n.col() },                               // Buffer dimensions
        { sizeof(double) * n.col(), sizeof(double) },       // Strides (in bytes) for each index
        n.addr()
    );
}, nullptr);
```
And then I wrote the following lines in the testing:
```
# with py::cast()
print(mat.array.base is mat)    # True
print(mat.array.flags)          # OWNDATA = FALSE

# without py::cast()
print(mat.array2.base is mat)   # False
print(mat.array2.flags)         # OWNDATA = TRUE
```

My understanding is that with `py::cast()`, it's like creating a view of a base object. In other words, they share the same buffer so the buffer content can be changed by `mat.array`. As for attributes like `shape` is managed separately by each object. More like a **shallow copy** of the base object.
But when a base object is not assigned, a **deep copy** of this object is created so that `mat.array2` cannot alter the content of `mat`'s buffer.

Maybe you can refer to this website: https://solegaonkar.github.io/NpArrayCopy.html

Reply:
It's no problem to use an analogy of shallow copy for the buffer sharing as long as you know that they are two difference things.